contracts/beyondFramework/dca/base/DCABaseUpgradeable.sol

```diff
@@ -0,0 +1,1168 @@
+import { IDcaStrategy } from "../interfaces/IDcaStrategy.sol";
+import { SwapLib } from "../libraries/SwapLib.sol";
+
+import { IERC20Upgradeable } from "@openzeppelin/contracts-upgradeable/interfaces/IERC20Upgradeable.sol";
+import { SafeERC20Upgradeable } from "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { PausableUpgradeable } from "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+
+abstract contract DCABaseUpgradeable is
+    OwnableUpgradeable,
+    PausableUpgradeable,
+    ReentrancyGuardUpgradeable,
+    IDcaStrategy
+{
+    using InvestQueueLib for InvestQueueLib.InvestQueue;
+    using DcaHistoryLib for DcaHistoryLib.DcaHistory;
+    using SwapLib for SwapLib.Router;
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    DepositFee public depositFee;
+
+    address public dcaInvestor;
+    address[] public portfolios;
+
+    TokenInfo public depositTokenInfo;
+    uint256 private depositTokenScale;
+
+    uint256 public investmentPeriod;
+    uint256 public lastInvestmentTimestamp;
+    uint256 public minDepositAmount;
+
+    uint16 public positionsLimit;
+
+    address[] public depositToBluechipSwapPath;
+    address[] public bluechipToDepositSwapPath;
+
+    BluechipInvestmentState public bluechipInvestmentState;
+
+    InvestQueueLib.InvestQueue private globalInvestQueue;
+    DcaHistoryLib.DcaHistory private dcaHistory;
+    SwapLib.Router public router;
+
+    TokenInfo public emergencyExitDepositToken;
+    uint256 public emergencySellDepositPrice;
+    TokenInfo public emergencyExitBluechipToken;
+    uint256 public emergencySellBluechipPrice;
+
+    mapping(address => DcaDepositor) private depositors;
+
+    uint256[7] private __gap;
+
+    // solhint-disable-next-line
+    function __DCABaseUpgradeable_init(DcaStrategyInitArgs calldata args)
+        internal
+        onlyInitializing
+    {
+        __Ownable_init();
+        __Pausable_init();
+        __ReentrancyGuard_init();
+
+        setBluechipInvestmentState(BluechipInvestmentState.Investing);
+        setDepositFee(args.depositFee);
+        setDcaInvestor(args.dcaInvestor);
+        setDepositTokenInto(args.depositTokenInfo);
+        setInvestmentPeriod(args.investmentPeriod);
+        setLastInvestmentTimestamp(args.lastInvestmentTimestamp);
+        setMinDepositAmount(args.minDepositAmount);
+        setPositionsLimit(args.positionsLimit);
+        setRouter(args.router);
+        setSwapPath(
+            args.depositToBluechipSwapPath,
+            args.bluechipToDepositSwapPath
+        );
+    }
+
+    modifier onlyDcaInvestor() {
+        require(_msgSender() == dcaInvestor, "Unauthorized");
+        _;
+    }
+
+    modifier onlyPortfolio() {
```

Comment:
isn't it more clean?

    modifier onlyPortfolio() {
        bool authorized;
        for (uint256 i = 0; i < portfolios.length; i++) {
            if (portfolios[i] == _msgSender()) {
                _;
                return;
            }
        }
require(authorized, "Unauthorized");	
       
    }